crates/driver/src/settlement_proposal.rs

```diff
@@ -0,0 +1,356 @@
+pub struct InteractionMetadata {
+    inputs: Vec<TokenAmount>,
+    outputs: Vec<TokenAmount>,
+    gas_used: U256,
+}
+
+/// A type which encodes the side effects of an on-chain interaction without necesserily containing
+/// the finalized call data. Is able to produce call data when it's required.
+#[async_trait::async_trait]
+#[cfg_attr(test, mockall::automock)]
+pub trait InteractionProposal: std::fmt::Debug + Send + Sync {
+    /// Turns something like an indicative RFQ interaction into a simulatable `Interaction` by
+    /// finalizing the required call data.
+    async fn finalize(&self) -> Result<Arc<dyn Interaction>>;
+
+    fn metadata(&self) -> InteractionMetadata;
+}
+
+#[derive(Debug, Clone)]
+pub struct TradedOrder {
+    pub order: Order,
+    pub executed_amount: U256,
+}
+
+impl TradedOrder {
+    /// Computes buy_token_price of the order based on whether it's a user order or a liquidity order.
+    /// User orders are allowed to get surplus and therefore return the clearing price of the
+    /// buy_token whereas liquidity orders must not get surplus so they return their limit price.
+    fn buy_token_price(&self, clearing_prices: &HashMap<H160, U256>) -> Option<U256> {
+        match self.order.metadata.is_liquidity_order {
+            // liquidity orders have to be settled at their limit price
+            true => clearing_prices
+                .get(&self.order.data.sell_token)?
+                .checked_mul(self.order.data.sell_amount)?
+                .checked_div(self.order.data.buy_amount),
+            false => clearing_prices.get(&self.order.data.buy_token).cloned(),
+        }
+    }
+
+    /// Computes and returns the executed trade amounts given the settlements uniform clearing
+    /// prices. The computed fee will always be 0 because it does not matter for further
+    /// computations.
+    fn execution(&self, clearing_prices: &HashMap<H160, U256>) -> Option<TradeExecution> {
+        let sell_price = clearing_prices.get(&self.order.data.sell_token)?;
+        let buy_price = self.buy_token_price(clearing_prices)?;
+
+        let order = &self.order.data;
+        let (sell_amount, buy_amount) = match order.kind {
+            OrderKind::Sell => {
+                let sell_amount = self.executed_amount;
+                let buy_amount = sell_amount
+                    .checked_mul(*sell_price)?
+                    .checked_ceil_div(&buy_price)?;
+                (sell_amount, buy_amount)
+            }
+            OrderKind::Buy => {
+                let buy_amount = self.executed_amount;
+                let sell_amount = buy_amount
+                    .checked_mul(buy_price)?
+                    .checked_div(*sell_price)?;
+                (sell_amount, buy_amount)
+            }
+        };
+
+        Some(TradeExecution {
+            sell_token: order.sell_token,
+            buy_token: order.buy_token,
+            sell_amount,
+            buy_amount,
+            fee_amount: 0.into(),
```

Comment:
Assuming this is still `TODO`.

Reply:
It's actually not. I mentioned that in the doc comment for the function.
`SettlementProposal::execution()` is only a private helper function and I think the `fee_amount` will not be relevant for further calculations if we base the ranking off of the surplus.

Reply:
I missed it sorry.

Reply:
Thinking about this more, we should probably return the fee amount as well because:
- The fee amount also needs to be added to the balances in `into_settlement_summary`
- The executed fee amount != fee amount (for partially fillable orders, since they are scaled based on how much is executed).